### PR TITLE
JBPM-10059 Aborted staged remains active in the process engine

### PR DESCRIPTION
**Pull request by bxf12315**
**Thank you for submitting this pull request**

**JIRA**: _(please edit the JIRA link if it exists)_ 

[https://issues.redhat.com/browse/JBPM-10059](https://issues.redhat.com/browse/JBPM-10059)

**referenced Pull Requests**: _(please edit the URLs of referenced pullrequests if they exist)_

* paste the link(s) from GitHub here
* link 2
* link 3 etc.

<details>
<summary>
How to replicate CI configuration locally?
</summary>

Build Chain tool does "simple" maven build(s), the builds are just Maven commands, but because the repositories relates and depends on each other and any change in API or class method could affect several of those repositories there is a need to use [build-chain tool](https://github.com/kiegroup/github-action-build-chain) to handle cross repository builds and be sure that we always use latest version of the code for each repository.
 
[build-chain tool](https://github.com/kiegroup/github-action-build-chain) is a build tool which can be used on command line locally or in Github Actions workflow(s), in case you need to change multiple repositories and send multiple dependent pull requests related with a change you can easily reproduce the same build by executing it on Github hosted environment or locally in your development environment. See [local execution](https://github.com/kiegroup/github-action-build-chain#local-execution) details to get more information about it.
</details>

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* <b>a pull request</b> please add comment: <b>Jenkins retest this</b>
 
* <b>a full downstream build</b> please add comment: <b>Jenkins run fdb</b>
  
* <b>a compile downstream build</b> please  add comment: <b>Jenkins run cdb</b>

* <b>a full production downstream build</b> please add comment: <b>Jenkins execute product fdb</b>

* <b>an upstream build</b> please add comment: <b>Jenkins run upstream</b>
</details>

### CODE REVIEW

**Comment by bxf12315** [comment]
@elguardian it's a draft code, please check it.

File: jbpm-case-mgmt/jbpm-case-mgmt-impl/src/main/java/org/jbpm/casemgmt/impl/CaseRuntimeDataServiceImpl.java [hunk 1]
```diff
@@ -702,6 +702,11 @@ public CaseInstance getCaseInstanceById(String caseId) {
                 status = StageStatus.Completed;
             }
             Collection<org.jbpm.services.api.model.NodeInstanceDesc> activeNodes = getActiveNodesForCaseAndStage(caseId, n.getNodeId(), new QueryContext(0, 100));
+
+            if (activeNodes == null || activeNodes.size() == 0) {
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-case-mgmt/jbpm-case-mgmt-impl/src/main/java/org/jbpm/casemgmt/impl/CaseRuntimeDataServiceImpl.java]: I meant the other way round... not having active nodes does not mean is completed.

[the remark this answers is not in the document]
Comment (bxf12315) [reply jbpm-case-mgmt/jbpm-case-mgmt-impl/src/main/java/org/jbpm/casemgmt/impl/CaseRuntimeDataServiceImpl.java]: new condition is if ((activeNodes == null || activeNodes.size() == 0) && ((CaseStageImpl) stagesByName.get(n.getNodeId())).isAutoComplete())

[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-case-mgmt/jbpm-case-mgmt-impl/src/main/java/org/jbpm/casemgmt/impl/CaseRuntimeDataServiceImpl.java]: you need to compute if a node is being completed or aborted. So far we have no way to know when a stage is being cancelled so my advice would be to check this line

'''
Collection<Long> completedNodes = nodes.stream().filter(n -> ((NodeInstanceDesc)n).getType() == 1).map(n -> n.getId()).collect(toList());
'''

and try to check if there are nodes in a different state as completed for instance (cancelled) if that happens you can see if you can set the status as cancelled or not

**Comment by gmunozfe** [comment]
@bxf12315 test failures seem related

**Comment by bxf12315** [comment]
> @bxf12315 test failures seem related

WDYT?  if the way is. correct,  I will add some UT for it.

**Review by gmunozfe — CHANGES_REQUESTED** [review]
@bxf12315 I've added some comments and also a test for this scenario, you can check it out here:
https://github.com/gmunozfe/jbpm/tree/JBPM-10059_test

I'm not sure what is the expected behaviour when a Stage is cancelled. According to this implementation (and based on autocomplete flag), the StageStatus is Completed, however, it seems more accurate to have Canceled as StageStatus, no matter if it has "autocomplete" or not. @elguardian wdyt?

Besides, it doesn't seem a valid fix, as this doesn't cover those without autocomplete.

File: jbpm-case-mgmt/jbpm-case-mgmt-impl/src/main/java/org/jbpm/casemgmt/impl/CaseRuntimeDataServiceImpl.java [hunk 2]
```diff
@@ -793,10 +798,14 @@ private Collection<CaseStage> collectCaseStages(String deploymentId, String proc
...
-                
+
+                boolean isAutoComplete = false;
+                if (dynamicNode.getCompletionExpression() != null) {
+                    isAutoComplete = dynamicNode.getCompletionExpression().equals("autocomplete");
+                }
                 result.add(new CaseStageImpl((String) ((DynamicNode) node).getMetaData("UniqueId"),
                                              node.getName(),
-                                             adHocFragments));
+                                             adHocFragments, isAutoComplete));
```
[the remark this answers is not in the document]
Comment (bxf12315) [reply jbpm-case-mgmt/jbpm-case-mgmt-impl/src/main/java/org/jbpm/casemgmt/impl/CaseRuntimeDataServiceImpl.java:808]: I think maybe is not, you can see commit https://github.com/kiegroup/jbpm/commit/fa4c3e8d9e197ed5b6f073e6b4f866d5f5e939ec#diff-d8bc14ba30316578b3079d5469901b48ada98d8be18ddc7e2367f969bb60917aR44
now, the default value is true.

**Comment by afalhambra** [comment]
@bxf12315 can this be closed? I can see a pr already merged for the same: https://github.com/kiegroup/jbpm/pull/2172

**Comment by bxf12315** [comment]
sure

**Pull request closed without merging**